Client/Frontend/Browser/Playlist/PlaylistMediaInfo.swift

```diff
@@ -0,0 +1,510 @@
+            contentInfoRequest = nil
+            return
+        }
+        
+        dataRequests.removeAll(where: { loadingRequest == $0 })
+    }
+    
+    func processPendingRequests() {
+        let requestsFulfilled = Set<AVAssetResourceLoadingRequest>(self.dataRequests.compactMap {
+            $0.contentInformationRequest?.contentLength = Int64(self.data.count)
+            $0.contentInformationRequest?.isByteRangeAccessSupported = true
+            
+            if self.haveEnoughDataToFulfillRequest($0.dataRequest!) {
+                $0.finishLoading()
+                return $0
+            }
+            return nil
+        })
+        
+        self.dataRequests.removeAll(where: { requestsFulfilled.contains($0) })
+    }
+    
+    func haveEnoughDataToFulfillRequest(_ dataRequest: AVAssetResourceLoadingDataRequest) -> Bool {
+        let requestedOffset = Int(dataRequest.requestedOffset)
+        let currentOffset = Int(dataRequest.currentOffset)
+        let requestedLength = dataRequest.requestedLength
+        
+        if currentOffset <= self.data.count {
+            let bytesToRespond = min(self.data.count - currentOffset, requestedLength)
+            let data = self.data.subdata(in: Range(uncheckedBounds: (currentOffset, currentOffset + bytesToRespond)))
+            dataRequest.respond(with: data)
+            return self.data.count >= requestedLength + requestedOffset
+        }
+        
+        return false
+    }
+}
+
+extension MediaResourceManager {
+    public static func shimURL(_ url: URL) -> URL {
+        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        guard let scheme = components?.scheme else { return url }
+        components?.scheme = "brave-media-resource" + scheme
+        return components?.url ?? url
+    }
+    
+    public static func unShimURL(_ url: URL) -> URL {
+        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        guard let scheme = components?.scheme else { return url }
+        components?.scheme = scheme.replacingOccurrences(of: "brave-media-resource", with: "")
+        return components?.url ?? url
+    }
+    
+    public static func isShimmedURL(_ url: URL) -> Bool {
+        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        guard let scheme = components?.scheme else { return false }
+        return scheme.hasPrefix("brave-media-resource")
+    }
+    
+    // Would be nice if AVPlayer could detect the mime-type from the URL for my delegate without a head request..
+    // This function only exists because I can't figure out why videos from URLs don't play unless I explicitly specify a mime-type..
+    public static func canStreamURL(_ url: URL, _ completion: @escaping (Bool) -> Void) {
+        getMimeType(url) { mimeType in
+            completion(!mimeType.isEmpty)
+        }
+    }
+    
+    public static func getMimeType(_ url: URL, _ completion: @escaping (String) -> Void) {
```

Comment:
completion should be optional string? On error you're just completing with an empty string which is odd